### PR TITLE
Add no_std + alloc feature configuration

### LINKED ISSUE #79: no_std + alloc cargo feature configuration
I would like to use `bstr` to implement a [conventionally UTF-8 `String` data structure](https://ruby-doc.org/core-2.6.3/String.html) to be used in a [Rust implementation of Ruby](https://github.com/artichoke/artichoke).

As I've been implementing the core data structures in Artichoke, I've made as many of them as I can `#[no_std]` (for example, [this `Array` implementation built with `Vec`](https://github.com/artichoke/artichoke/blob/1f31c37d1762fe6c109525bd03d38f47bd214b3c/spinoso-array/src/lib.rs#L135-L136)).

The `ByteVec` extension trait has several methods I'd like to use instead of reimplementing myself, such as `ByteVec::remove_char`. Unfortunately, the `ByteVec` impl is gated on the `std` cargo feature which will require my crate to link to `std` despite my interest in only linking to `alloc`. For this `String` data structure, I am not interested in any of the `PathBuf` or `OsString` APIs.

I think this might be a breaking change for 1.0, so I wanted to raise this as soon as I ran into it.

### PR DESCRIPTION
**Pull request by lopopolo**
This commit enables `bstr` to build and test without a dependency on
`std`. This change was mostly munging `use` statements to prefer `core`
and `alloc` and changing around `cfg`s.

This change also enables `bstr` to successfully test in `no_std` and
`alloc` configurations. The following succeed where they did not even
compile before:

```
cargo test --no-default-features --lib
cargo test --no-default-features --lib --features alloc
```

Fixes #79.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -366,17 +366,20 @@ to write correct code for Unix, at the cost of getting a corner case wrong on
 Windows.
 */
 
-#![cfg_attr(not(feature = "std"), no_std)]
+#![no_std]
```
Comment (lopopolo) [inline src/lib.rs:369]: This change requires all items in the prelude to be imported even when linking `std`. It made fixing all of the import errors much easier. I also saw a recommendation in an issue on the Rust issue tracker that preferred this syntax. I think I've seen it used in `wasm-bindgen`.

**Comment by lopopolo** [comment]
I wanted to note that none of the examples compile without default features since they depend on the `std` feature. I think the way to fix this is add `[[example]]` stanzas to `Cargo.toml` with `required-features = "std"`, but I considered this too invasive for this PR so decided against doing it myself.

File: src/lib.rs [hunk 2]
```diff
@@ -366,17 +366,20 @@ to write correct code for Unix, at the cost of getting a corner case wrong on
...
 #![allow(dead_code)]
 
+#[cfg(feature = "alloc")]
+extern crate alloc;
+
 #[cfg(feature = "std")]
-extern crate core;
+extern crate std;
```
Comment (lopopolo) [inline src/lib.rs:376]: If you'd prefer, I can make this `#[cfg(any(test, feature = "std"))]` and remove the `alloc` gates from tests.

**Comment by programmerjake** [comment]
ping -- I'd like to use bstr with alloc but without std -- I'm trying to use BString kinda like OsString as the backing type for a POSIX Path type for a virtualized filesystem I'm writing.

**Comment by lopopolo** [comment]
@BurntSushi Friendly poke. Are you interested in this patch?

File: src/lib.rs (hunk 2 above)
Comment (BurntSushi) [reply src/lib.rs:376]: Yes, I would like that. Otherwise, I think writing tests becomes too annoying. With that said, I do think that things like `cargo t --no-default-features` and `cargo t --no-default-features --features alloc` should at least compile fine (and run any relevant tests, if any, probably none). So I think most tests do need to be gated on `#[cfg(feature = "std")]`.

File: src/lib.rs (hunk 1 above)
Comment (BurntSushi) [reply src/lib.rs:369]: Hmmm, this is not how I've done it in the past. Could you say more about what becomes easier with this?

One thing that this makes kind of a pain is that (AIUI) we can't benefit from std's prelude in tests if we do this, right? That's kind of a bummer and I'd rather have the prelude available in tests.

**Review by BurntSushi — CHANGES_REQUESTED** [review]
Thanks! I left a couple of minor comments.

I'm of course interested in this PR, but me taking months to get bigger PRs merged is unfortunately just reality. I maintain a lot of projects and it is very costly for me to be constantly context switching.

One thing I noticed in this diff is that `use` statements have gotten a bit messier. Ideally, we'd switch over to nested `use` statements, but I think this diff is big enough already. With that said, one improvement would be to consistently keep `use` statements in at least three distinct blocks: `core/alloc/std`, then third-party crate imports and then finally crate-internal imports. In some parts of this PR, I see the first two categories lumped together.

I think some kind of docs are also needed here. Probably at least in the README where the crate features are documented.

File: src/lib.rs (hunk 1 above)
Comment (lopopolo) [reply src/lib.rs]: I found the docs that refer to this approach:

https://github.com/rust-lang/cargo/blob/58a961314437258065e23cb6316dfc121d96fb71/src/doc/src/reference/features-examples.md#no_std-support

Comment (thomcc) [reply src/lib.rs]: > One thing that this makes kind of a pain is that (AIUI) we can't benefit from std's prelude in tests if we do this, right? That's kind of a bummer and I'd rather have the prelude available in tests.

You can do this too:
```
#![cfg_attr(not(test), no_std)]
#[cfg(feature = "std")]
extern crate std;
```

(or `#[cfg(all(not(test), feature = "std"))]` on the `extern crate std` — I don't think there's a difference)

But given that bstr already has most of the tests in `mod tests` you could just put `use std::prelude::v1::*` at the top of it and get most of the benefit.

It's kind of annoying that there's not a better way to do this — the way the prelude works for no_std is janky, honestly.

**Comment by lopopolo** [comment]
hey @BurntSushi, I know you have a lot of surface area on GitHub. This is a friendly poke that I've addressed all of your feedback on this PR and rebased it to the latest master.
